### PR TITLE
Docker build files

### LINKED ISSUE #125: Publish release to docker
It would be useful to have releases published to docker. This would allow easy testing in environments like Kubernetes, or where I don't have bazel, etc

### PR DESCRIPTION
**Pull request by oschaaf**
Based on #211 by @howardjohn (thanks!), this adds basic docker files for
all the NH binaries plus a CI docker build test.

Note this PR deprecates the explicit opt build in CI at the same time,
as that now is implicitly tested via the Docker CI task.

Fixes #125
Upon merge we can close #211

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
(force pushed a rebase for fixing DCO)

**Comment by oschaaf** [comment]
thanks for taking a look @howardjohn !

File: ci/Dockerfile-nighthawk-test-server [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM frolvlad/alpine-glibc
```
Comment (htuch) [inline ci/Dockerfile-nighthawk-test-server]: Suggest moving these to their own dedicated directory, e.g. `ci/docker`, `ci/images` or something

File: ci/Dockerfile-nighthawk-output-transform [hunk 2]
```diff
@@ -0,0 +1,5 @@
+FROM frolvlad/alpine-glibc
```
Comment (htuch) [inline ci/Dockerfile-nighthawk-output-transform]: Do we really want a separate image for each tool? Can't we just put them in one? I imagine the size difference isn't significant.

Comment (oschaaf) [reply ci/Dockerfile-nighthawk-output-transform]: +1, pushed a change that uses a single image in f6478cea7880a077e441d86b017c053d6efcb7f4
The binary sizes look rather large though, looking into that.
@howardjohn binary size aside, any objections to this moving all the tools into a single image?

Comment (howardjohn) [reply ci/Dockerfile-nighthawk-output-transform]: Nope, seems like a good idea to me

Comment (howardjohn) [reply ci/Dockerfile-nighthawk-output-transform]: Sounds good to me

Comment (oschaaf) [reply ci/Dockerfile-nighthawk-output-transform]: great, added 1a30e966748f1da530f7b17880b52f1e5c17b7a6 to strip the binaries

**Comment by oschaaf** [comment]
Also /assign jplevyak

File: ci/Dockerfile-nighthawk-test-server (hunk 1 above)
Comment (oschaaf) [reply ci/Dockerfile-nighthawk-test-server]: Done

**Comment by oschaaf** [comment]
remaining comment resolved in a55fed9e3d5c27df742b024891d9899c10324f70, plus:

- remove -x. see code-level comment about why 
- add some top level comments
- use a working directory for the build context
- make docker_build.sh less picky wrt the location it is run from
- use bazel info to query paths

File: ci/docker/docker_build.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# NOTE: explicit no -x for verbose commands. Because this is run in CI, doing so may result in
+# publishing sensitive information into public CI logs if someone makes a change in a
+# consuming script that is off guard.
+
+DOCKER_NAME="nighthawk"
+DOCKER_IMAGE_PREFIX="envoyproxy/${DOCKER_NAME}"
+BINARIES=(nighthawk_test_server nighthawk_client nighthawk_service nighthawk_output_transform)
+BAZEL_BIN=$(bazel info -c opt bazel-bin)
+WORKSPACE=$(bazel info workspace)
+TMP_DIR=${WORKSPACE}/tmp-docker-build-context
```
Comment (htuch) [inline ci/docker/docker_build.sh]: Nit: please `""` wrap all the expansions in this file for whitespace safety.

Comment (oschaaf) [reply ci/docker/docker_build.sh]: Done in 334fa1281e43f78b5b4a1cec6052d7ab98a5423e

**Pull request merged**